File: ios/Runner/P2PHandler.swift
import Foundation
import MultipeerConnectivity

/// Handles MultipeerConnectivity P2P communication for YourTurn app.
/// Manages session advertising, browsing, and message passing.
class P2PHandler: NSObject {

    // MARK: - Constants
    private static let serviceType = "yourturn-game" // Max 15 chars, lowercase alphanumeric and hyphens

    // MARK: - Properties
    private var peerID: MCPeerID!
    private var session: MCSession?
    private var advertiser: MCNearbyServiceAdvertiser?
    private var browser: MCNearbyServiceBrowser?

    private var currentSessionId: String?
    private var currentSessionCode: String?
    private var isHost: Bool = false
    private var seqNo: Int = 0

    // Event handlers (set by AppDelegate)
    var onSessionDiscovered: (([String: Any]) -> Void)?
    var onPeerConnected: ((String, String) -> Void)?
    var onPeerDisconnected: ((String) -> Void)?
    var onMessageReceived: (([String: Any]) -> Void)?
    var onError: ((String) -> Void)?

    // Track discovered sessions
    private var discoveredPeers: [MCPeerID: [String: String]] = [:]

    // Track connected players
    private var connectedPlayers: [MCPeerID: [String: Any]] = [:]

    // MARK: - Initialization
    override init() {
        super.init()
    }

    // MARK: - Public Methods

    /// Create a new session as host
    func createSession(leaderName: String, sessionId: String, sessionCode: String) -> [String: Any] {
        cleanup()

        peerID = MCPeerID(displayName: leaderName)
        session = MCSession(peer: peerID, securityIdentity: nil, encryptionPreference: .required)
        session?.delegate = self

        currentSessionId = sessionId
        currentSessionCode = sessionCode
        isHost = true
        seqNo = 0

        // Start advertising with session info
        let discoveryInfo: [String: String] = [
            "sessionId": sessionId,
            "sessionCode": sessionCode,
            "leaderName": leaderName,
            "isInProgress": "false"
        ]

        advertiser = MCNearbyServiceAdvertiser(
            peer: peerID,
            discoveryInfo: discoveryInfo,
            serviceType: P2PHandler.serviceType
        )
        advertiser?.delegate = self
        advertiser?.startAdvertisingPeer()

        return [
            "sessionId": sessionId,
            "sessionCode": sessionCode,
            "leaderId": peerID.displayName,
            "success": true
        ]
    }

    /// Start browsing for nearby sessions
    func startDiscovery() {
        if peerID == nil {
            peerID = MCPeerID(displayName: UUID().uuidString.prefix(8).description)
        }

        browser?.stopBrowsingForPeers()
        browser = MCNearbyServiceBrowser(peer: peerID, serviceType: P2PHandler.serviceType)
        browser?.delegate = self
        browser?.startBrowsingForPeers()
    }

    /// Stop browsing for sessions
    func stopDiscovery() {
        browser?.stopBrowsingForPeers()
        browser = nil
    }

    /// Join an existing session
    func joinSession(sessionCode: String, playerName: String, completion: @escaping (Result<[String: Any], Error>) -> Void) {
        // Find the peer with matching session code
        guard let (hostPeer, info) = discoveredPeers.first(where: { $0.value["sessionCode"] == sessionCode }) else {
            completion(.failure(NSError(domain: "P2P", code: 404, userInfo: [NSLocalizedDescriptionKey: "Session not found"])))
            return
        }

        cleanup()

        peerID = MCPeerID(displayName: playerName)
        session = MCSession(peer: peerID, securityIdentity: nil, encryptionPreference: .required)
        session?.delegate = self

        currentSessionId = info["sessionId"]
        currentSessionCode = sessionCode
        isHost = false

        // Store completion to call when connected
        pendingJoinCompletion = completion
        pendingJoinHostPeer = hostPeer

        // Create browser to invite ourselves
        browser = MCNearbyServiceBrowser(peer: peerID, serviceType: P2PHandler.serviceType)
        browser?.delegate = self
        browser?.startBrowsingForPeers()

        // Invite the host
        if let session = session {
            browser?.invitePeer(hostPeer, to: session, withContext: playerName.data(using: .utf8), timeout: 30)
        }
    }

    private var pendingJoinCompletion: ((Result<[String: Any], Error>) -> Void)?
    private var pendingJoinHostPeer: MCPeerID?

    /// Start the game (update advertising info)
    func startGame() {
        guard isHost else { return }

        advertiser?.stopAdvertisingPeer()

        let discoveryInfo: [String: String] = [
            "sessionId": currentSessionId ?? "",
            "sessionCode": currentSessionCode ?? "",
            "leaderName": peerID.displayName,
            "isInProgress": "true"
        ]

        advertiser = MCNearbyServiceAdvertiser(
            peer: peerID,
            discoveryInfo: discoveryInfo,
            serviceType: P2PHandler.serviceType
        )
        advertiser?.delegate = self
        advertiser?.startAdvertisingPeer()

        // Broadcast game start to all peers
        broadcast(message: [
            "type": "gameStart",
            "sessionId": currentSessionId ?? "",
            "seqNo": incrementSeqNo()
        ])
    }

    /// End the game
    func endGame() {
        broadcast(message: [
            "type": "gameEnd",
            "sessionId": currentSessionId ?? "",
            "seqNo": incrementSeqNo()
        ])
    }

    /// Pass turn to next player
    func passTurn(toPlayerId: String, fromPlayerId: String) {
        broadcast(message: [
            "type": "turnChange",
            "sessionId": currentSessionId ?? "",
            "toPlayerId": toPlayerId,
            "fromPlayerId": fromPlayerId,
            "seqNo": incrementSeqNo()
        ])
    }

    /// Update timer settings
    func updateTimerSetting(minutes: Int?) {
        broadcast(message: [
            "type": "timerUpdate",
            "sessionId": currentSessionId ?? "",
            "minutes": minutes as Any,
            "seqNo": incrementSeqNo()
        ])
    }

    /// Update start player
    func updateStartPlayer(index: Int) {
        broadcast(message: [
            "type": "startPlayerUpdate",
            "sessionId": currentSessionId ?? "",
            "startPlayerIndex": index,
            "seqNo": incrementSeqNo()
        ])
    }

    /// Reorder players
    func reorderPlayers(playerIds: [String]) {
        broadcast(message: [
            "type": "playersReorder",
            "sessionId": currentSessionId ?? "",
            "playerIds": playerIds,
            "seqNo": incrementSeqNo()
        ])
    }

    /// Leave the current session
    func leaveSession() {
        broadcast(message: [
            "type": "playerLeft",
            "sessionId": currentSessionId ?? "",
            "playerId": peerID?.displayName ?? "",
            "seqNo": incrementSeqNo()
        ])
        cleanup()
    }

    /// Clean up all resources
    func cleanup() {
        advertiser?.stopAdvertisingPeer()
        advertiser = nil
        browser?.stopBrowsingForPeers()
        browser = nil
        session?.disconnect()
        session = nil
        currentSessionId = nil
        currentSessionCode = nil
        isHost = false
        connectedPlayers.removeAll()
    }

    // MARK: - Private Methods

    private func incrementSeqNo() -> Int {
        seqNo += 1
        return seqNo
    }

    private func broadcast(message: [String: Any]) {
        guard let session = session, !session.connectedPeers.isEmpty else { return }

        do {
            let data = try JSONSerialization.data(withJSONObject: message)
            try session.send(data, toPeers: session.connectedPeers, with: .reliable)
        } catch {
            onError?("Failed to broadcast: \(error.localizedDescription)")
        }
    }

    private func send(message: [String: Any], to peer: MCPeerID) {
        guard let session = session else { return }

        do {
            let data = try JSONSerialization.data(withJSONObject: message)
            try session.send(data, toPeers: [peer], with: .reliable)
        } catch {
            onError?("Failed to send: \(error.localizedDescription)")
        }
    }
}

// MARK: - MCSessionDelegate
extension P2PHandler: MCSessionDelegate {
    func session(_ session: MCSession, peer peerID: MCPeerID, didChange state: MCSessionState) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            switch state {
            case .connected:
                print("[P2P] Peer connected: \(peerID.displayName)")

                // If we're joining and this is the host
                if let completion = self.pendingJoinCompletion,
                   peerID == self.pendingJoinHostPeer {
                    self.pendingJoinCompletion = nil
                    self.pendingJoinHostPeer = nil

                    completion(.success([
                        "sessionId": self.currentSessionId ?? "",
                        "sessionCode": self.currentSessionCode ?? "",
                        "playerId": self.peerID.displayName,
                        "success": true
                    ]))
                }

                // Notify about connection
                self.onPeerConnected?(peerID.displayName, self.currentSessionId ?? "")

                // If we're host, send current state to new peer
                if self.isHost {
                    self.sendCurrentState(to: peerID)
                }

            case .notConnected:
                print("[P2P] Peer disconnected: \(peerID.displayName)")
                self.connectedPlayers.removeValue(forKey: peerID)
                self.onPeerDisconnected?(peerID.displayName)

            case .connecting:
                print("[P2P] Connecting to peer: \(peerID.displayName)")

            @unknown default:
                break
            }
        }
    }

    func session(_ session: MCSession, didReceive data: Data, fromPeer peerID: MCPeerID) {
        guard let message = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }

        DispatchQueue.main.async { [weak self] in
            self?.onMessageReceived?(message)
        }
    }

    func session(_ session: MCSession, didReceive stream: InputStream, withName streamName: String, fromPeer peerID: MCPeerID) {
        // Not used
    }

    func session(_ session: MCSession, didStartReceivingResourceWithName resourceName: String, fromPeer peerID: MCPeerID, with progress: Progress) {
        // Not used
    }

    func session(_ session: MCSession, didFinishReceivingResourceWithName resourceName: String, fromPeer peerID: MCPeerID, at localURL: URL?, withError error: Error?) {
        // Not used
    }

    private func sendCurrentState(to peer: MCPeerID) {
        // Host sends current session state to newly connected peer
        let connectedPlayerNames = (session?.connectedPeers.map { $0.displayName } ?? []) + [peerID.displayName]

        send(message: [
            "type": "sessionState",
            "sessionId": currentSessionId ?? "",
            "players": connectedPlayerNames,
            "seqNo": seqNo
        ], to: peer)
    }
}

// MARK: - MCNearbyServiceAdvertiserDelegate
extension P2PHandler: MCNearbyServiceAdvertiserDelegate {
    func advertiser(_ advertiser: MCNearbyServiceAdvertiser, didReceiveInvitationFromPeer peerID: MCPeerID, withContext context: Data?, invitationHandler: @escaping (Bool, MCSession?) -> Void) {
        // Accept all invitations (players joining the session)
        print("[P2P] Received invitation from: \(peerID.displayName)")

        // Store player info from context
        if let context = context, let playerName = String(data: context, encoding: .utf8) {
            connectedPlayers[peerID] = ["name": playerName, "id": peerID.displayName]
        }

        invitationHandler(true, session)
    }

    func advertiser(_ advertiser: MCNearbyServiceAdvertiser, didNotStartAdvertisingPeer error: Error) {
        onError?("Failed to start advertising: \(error.localizedDescription)")
    }
}

// MARK: - MCNearbyServiceBrowserDelegate
extension P2PHandler: MCNearbyServiceBrowserDelegate {
    func browser(_ browser: MCNearbyServiceBrowser, foundPeer peerID: MCPeerID, withDiscoveryInfo info: [String : String]?) {
        print("[P2P] Found peer: \(peerID.displayName) with info: \(info ?? [:])")

        if let info = info {
            discoveredPeers[peerID] = info

            // Notify Flutter about discovered session
            DispatchQueue.main.async { [weak self] in
                self?.onSessionDiscovered?([
                    "code": info["sessionCode"] ?? "",
                    "advertisedBy": info["leaderName"] ?? peerID.displayName,
                    "isInProgress": info["isInProgress"] == "true"
                ])
            }
        }
    }

    func browser(_ browser: MCNearbyServiceBrowser, lostPeer peerID: MCPeerID) {
        print("[P2P] Lost peer: \(peerID.displayName)")
        discoveredPeers.removeValue(forKey: peerID)
    }

    func browser(_ browser: MCNearbyServiceBrowser, didNotStartBrowsingForPeers error: Error) {
        onError?("Failed to start browsing: \(error.localizedDescription)")
    }
}

File: android/app/src/main/kotlin/com/tendimensions/yourturn/P2PHandler.kt
package com.tendimensions.yourturn

import android.content.Context
import android.util.Log
import com.google.android.gms.nearby.Nearby
import com.google.android.gms.nearby.connection.*
import org.json.JSONArray
import org.json.JSONObject
import java.nio.charset.StandardCharsets

/**
 * Handles Nearby Connections P2P communication for YourTurn app.
 * Manages session advertising, discovery, and message passing.
 */
class P2PHandler(private val context: Context) {

    companion object {
        private const val TAG = "P2PHandler"
        private const val SERVICE_ID = "com.tendimensions.yourturn.p2p"
    }

    // Nearby Connections client
    private val connectionsClient: ConnectionsClient = Nearby.getConnectionsClient(context)

    // Session state
    private var currentSessionId: String? = null
    private var currentSessionCode: String? = null
    private var myPlayerName: String? = null
    private var isHost: Boolean = false
    private var seqNo: Int = 0

    // Connected endpoints (endpointId -> playerName)
    private val connectedEndpoints: MutableMap<String, String> = mutableMapOf()

    // Discovered sessions (endpointId -> session info)
    private val discoveredSessions: MutableMap<String, Map<String, String>> = mutableMapOf()

    // Pending join completion
    private var pendingJoinCallback: ((Result<Map<String, Any>>) -> Unit)? = null
    private var pendingJoinEndpointId: String? = null

    // Event handlers (set by MainActivity)
    var onSessionDiscovered: ((Map<String, Any>) -> Unit)? = null
    var onPeerConnected: ((String, String) -> Unit)? = null
    var onPeerDisconnected: ((String) -> Unit)? = null
    var onMessageReceived: ((Map<String, Any>) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    // Connection lifecycle callbacks
    private val connectionLifecycleCallback = object : ConnectionLifecycleCallback() {
        override fun onConnectionInitiated(endpointId: String, info: ConnectionInfo) {
            Log.d(TAG, "Connection initiated with: ${info.endpointName}")
            // Auto-accept all connections
            connectionsClient.acceptConnection(endpointId, payloadCallback)
        }

        override fun onConnectionResult(endpointId: String, result: ConnectionResolution) {
            when (result.status.statusCode) {
                ConnectionsStatusCodes.STATUS_OK -> {
                    Log.d(TAG, "Connected to endpoint: $endpointId")

                    // If we're joining and this is the host we connected to
                    if (pendingJoinCallback != null && pendingJoinEndpointId == endpointId) {
                        val callback = pendingJoinCallback
                        pendingJoinCallback = null
                        pendingJoinEndpointId = null

                        callback?.invoke(Result.success(mapOf(
                            "sessionId" to (currentSessionId ?: ""),
                            "sessionCode" to (currentSessionCode ?: ""),
                            "playerId" to (myPlayerName ?: ""),
                            "success" to true
                        )))
                    }

                    // Notify about connection
                    val playerName = connectedEndpoints[endpointId] ?: endpointId
                    onPeerConnected?.invoke(playerName, currentSessionId ?: "")

                    // If we're host, send current state to new peer
                    if (isHost) {
                        sendCurrentState(endpointId)
                    }
                }
                ConnectionsStatusCodes.STATUS_CONNECTION_REJECTED -> {
                    Log.d(TAG, "Connection rejected by: $endpointId")
                    pendingJoinCallback?.invoke(Result.failure(Exception("Connection rejected")))
                    pendingJoinCallback = null
                }
                ConnectionsStatusCodes.STATUS_ERROR -> {
                    Log.e(TAG, "Connection error with: $endpointId")
                    pendingJoinCallback?.invoke(Result.failure(Exception("Connection error")))
                    pendingJoinCallback = null
                }
            }
        }

        override fun onDisconnected(endpointId: String) {
            Log.d(TAG, "Disconnected from endpoint: $endpointId")
            val playerName = connectedEndpoints.remove(endpointId) ?: endpointId
            onPeerDisconnected?.invoke(playerName)
        }
    }

    // Payload callback for receiving messages
    private val payloadCallback = object : PayloadCallback() {
        override fun onPayloadReceived(endpointId: String, payload: Payload) {
            if (payload.type == Payload.Type.BYTES) {
                payload.asBytes()?.let { bytes ->
                    try {
                        val json = String(bytes, StandardCharsets.UTF_8)
                        val message = jsonToMap(JSONObject(json))
                        onMessageReceived?.invoke(message)
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to parse message", e)
                    }
                }
            }
        }

        override fun onPayloadTransferUpdate(endpointId: String, update: PayloadTransferUpdate) {
            // Not needed for byte payloads
        }
    }

    // Endpoint discovery callback
    private val endpointDiscoveryCallback = object : EndpointDiscoveryCallback() {
        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
            Log.d(TAG, "Found endpoint: $endpointId, name: ${info.endpointName}")

            // Parse endpoint name as JSON containing session info
            try {
                val sessionInfo = JSONObject(info.endpointName)
                val infoMap = mapOf(
                    "sessionId" to sessionInfo.optString("sessionId", ""),
                    "sessionCode" to sessionInfo.optString("sessionCode", ""),
                    "leaderName" to sessionInfo.optString("leaderName", ""),
                    "isInProgress" to sessionInfo.optString("isInProgress", "false")
                )
                discoveredSessions[endpointId] = infoMap

                onSessionDiscovered?.invoke(mapOf(
                    "code" to infoMap["sessionCode"]!!,
                    "advertisedBy" to infoMap["leaderName"]!!,
                    "isInProgress" to (infoMap["isInProgress"] == "true")
                ))
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse endpoint info", e)
            }
        }

        override fun onEndpointLost(endpointId: String) {
            Log.d(TAG, "Lost endpoint: $endpointId")
            discoveredSessions.remove(endpointId)
        }
    }

    // MARK: - Public Methods

    /**
     * Create a new session as host
     */
    fun createSession(leaderName: String, sessionId: String, sessionCode: String): Map<String, Any> {
        cleanup()

        myPlayerName = leaderName
        currentSessionId = sessionId
        currentSessionCode = sessionCode
        isHost = true
        seqNo = 0

        // Create endpoint name as JSON with session info
        val endpointName = JSONObject().apply {
            put("sessionId", sessionId)
            put("sessionCode", sessionCode)
            put("leaderName", leaderName)
            put("isInProgress", "false")
        }.toString()

        // Start advertising
        val advertisingOptions = AdvertisingOptions.Builder()
            .setStrategy(Strategy.P2P_STAR)
            .build()

        connectionsClient.startAdvertising(
            endpointName,
            SERVICE_ID,
            connectionLifecycleCallback,
            advertisingOptions
        ).addOnSuccessListener {
            Log.d(TAG, "Started advertising session: $sessionCode")
        }.addOnFailureListener { e ->
            Log.e(TAG, "Failed to start advertising", e)
            onError?.invoke("Failed to start advertising: ${e.message}")
        }

        return mapOf(
            "sessionId" to sessionId,
            "sessionCode" to sessionCode,
            "leaderId" to leaderName,
            "success" to true
        )
    }

    /**
     * Start discovering nearby sessions
     */
    fun startDiscovery() {
        val discoveryOptions = DiscoveryOptions.Builder()
            .setStrategy(Strategy.P2P_STAR)
            .build()

        connectionsClient.startDiscovery(
            SERVICE_ID,
            endpointDiscoveryCallback,
            discoveryOptions
        ).addOnSuccessListener {
            Log.d(TAG, "Started discovery")
        }.addOnFailureListener { e ->
            Log.e(TAG, "Failed to start discovery", e)
            onError?.invoke("Failed to start discovery: ${e.message}")
        }
    }

    /**
     * Stop discovering sessions
     */
    fun stopDiscovery() {
        connectionsClient.stopDiscovery()
    }

    /**
     * Join an existing session by code
     */
    fun joinSession(sessionCode: String, playerName: String, callback: (Result<Map<String, Any>>) -> Unit) {
        // Find the endpoint with matching session code
        val entry = discoveredSessions.entries.find { it.value["sessionCode"] == sessionCode }

        if (entry == null) {
            callback(Result.failure(Exception("Session not found")))
            return
        }

        val (endpointId, info) = entry

        cleanup()

        myPlayerName = playerName
        currentSessionId = info["sessionId"]
        currentSessionCode = sessionCode
        isHost = false

        pendingJoinCallback = callback
        pendingJoinEndpointId = endpointId

        // Store the endpoint as a connected player
        connectedEndpoints[endpointId] = info["leaderName"] ?: "Host"

        // Request connection to the host
        connectionsClient.requestConnection(
            playerName,
            endpointId,
            connectionLifecycleCallback
        ).addOnFailureListener { e ->
            Log.e(TAG, "Failed to request connection", e)
            pendingJoinCallback?.invoke(Result.failure(e))
            pendingJoinCallback = null
        }
    }

    /**
     * Start the game (update advertising info)
     */
    fun startGame() {
        if (!isHost) return

        // Stop and restart advertising with updated info
        connectionsClient.stopAdvertising()

        val endpointName = JSONObject().apply {
            put("sessionId", currentSessionId ?: "")
            put("sessionCode", currentSessionCode ?: "")
            put("leaderName", myPlayerName ?: "")
            put("isInProgress", "true")
        }.toString()

        val advertisingOptions = AdvertisingOptions.Builder()
            .setStrategy(Strategy.P2P_STAR)
            .build()

        connectionsClient.startAdvertising(
            endpointName,
            SERVICE_ID,
            connectionLifecycleCallback,
            advertisingOptions
        )

        // Broadcast game start to all peers
        broadcast(mapOf(
            "type" to "gameStart",
            "sessionId" to (currentSessionId ?: ""),
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * End the game
     */
    fun endGame() {
        broadcast(mapOf(
            "type" to "gameEnd",
            "sessionId" to (currentSessionId ?: ""),
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * Pass turn to next player
     */
    fun passTurn(toPlayerId: String, fromPlayerId: String) {
        broadcast(mapOf(
            "type" to "turnChange",
            "sessionId" to (currentSessionId ?: ""),
            "toPlayerId" to toPlayerId,
            "fromPlayerId" to fromPlayerId,
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * Update timer settings
     */
    fun updateTimerSetting(minutes: Int?) {
        broadcast(mapOf(
            "type" to "timerUpdate",
            "sessionId" to (currentSessionId ?: ""),
            "minutes" to (minutes ?: JSONObject.NULL),
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * Update start player
     */
    fun updateStartPlayer(index: Int) {
        broadcast(mapOf(
            "type" to "startPlayerUpdate",
            "sessionId" to (currentSessionId ?: ""),
            "startPlayerIndex" to index,
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * Reorder players
     */
    fun reorderPlayers(playerIds: List<String>) {
        broadcast(mapOf(
            "type" to "playersReorder",
            "sessionId" to (currentSessionId ?: ""),
            "playerIds" to playerIds,
            "seqNo" to incrementSeqNo()
        ))
    }

    /**
     * Leave the current session
     */
    fun leaveSession() {
        broadcast(mapOf(
            "type" to "playerLeft",
            "sessionId" to (currentSessionId ?: ""),
            "playerId" to (myPlayerName ?: ""),
            "seqNo" to incrementSeqNo()
        ))
        cleanup()
    }

    /**
     * Clean up all resources
     */
    fun cleanup() {
        connectionsClient.stopAdvertising()
        connectionsClient.stopDiscovery()
        connectionsClient.stopAllEndpoints()

        connectedEndpoints.clear()
        discoveredSessions.clear()
        currentSessionId = null
        currentSessionCode = null
        isHost = false
        pendingJoinCallback = null
        pendingJoinEndpointId = null
    }

    // MARK: - Private Methods

    private fun incrementSeqNo(): Int {
        seqNo += 1
        return seqNo
    }

    private fun broadcast(message: Map<String, Any>) {
        if (connectedEndpoints.isEmpty()) return

        try {
            val json = mapToJson(message).toString()
            val payload = Payload.fromBytes(json.toByteArray(StandardCharsets.UTF_8))

            for (endpointId in connectedEndpoints.keys) {
                connectionsClient.sendPayload(endpointId, payload)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to broadcast message", e)
            onError?.invoke("Failed to broadcast: ${e.message}")
        }
    }

    private fun sendToEndpoint(endpointId: String, message: Map<String, Any>) {
        try {
            val json = mapToJson(message).toString()
            val payload = Payload.fromBytes(json.toByteArray(StandardCharsets.UTF_8))
            connectionsClient.sendPayload(endpointId, payload)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send message", e)
            onError?.invoke("Failed to send: ${e.message}")
        }
    }

    private fun sendCurrentState(endpointId: String) {
        // Host sends current session state to newly connected peer
        val playerNames = connectedEndpoints.values.toMutableList()
        playerNames.add(0, myPlayerName ?: "")

        sendToEndpoint(endpointId, mapOf(
            "type" to "sessionState",
            "sessionId" to (currentSessionId ?: ""),
            "players" to playerNames,
            "seqNo" to seqNo
        ))
    }

    // JSON conversion helpers
    private fun mapToJson(map: Map<String, Any>): JSONObject {
        val json = JSONObject()
        for ((key, value) in map) {
            when (value) {
                is List<*> -> json.put(key, JSONArray(value))
                is Map<*, *> -> json.put(key, mapToJson(value as Map<String, Any>))
                else -> json.put(key, value)
            }
        }
        return json
    }

    private fun jsonToMap(json: JSONObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        for (key in json.keys()) {
            val value = json.get(key)
            map[key] = when (value) {
                is JSONObject -> jsonToMap(value)
                is JSONArray -> jsonArrayToList(value)
                JSONObject.NULL -> ""
                else -> value
            }
        }
        return map
    }

    private fun jsonArrayToList(array: JSONArray): List<Any> {
        val list = mutableListOf<Any>()
        for (i in 0 until array.length()) {
            val value = array.get(i)
            list.add(when (value) {
                is JSONObject -> jsonToMap(value)
                is JSONArray -> jsonArrayToList(value)
                else -> value
            })
        }
        return list
    }
}
